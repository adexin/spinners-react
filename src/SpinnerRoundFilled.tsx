import React from 'react';
import { defaultProps, SpinnerProps } from './helpers';
import { SpinnersProps, withSharedProps } from './withSharedProps';

import './SpinnerRoundFilled.css';

const animations = [
  { r: 4 },
  {
    name: 'spinners-react-round-filled-inner',
    r: 12.66,
  },
  {
    name: 'spinners-react-round-filled-center',
    r: 20.32,
  },
  {
    name: 'spinners-react-round-filled-outer',
    r: 27.5,
  },
];

export type SpinnerRoundFilledProps = SpinnersProps & SpinnerProps;

const Component = ({
  speed = defaultProps.speed,
  still = defaultProps.still,
  thickness = defaultProps.thickness,
  ...svgProps
}: SpinnerProps) => (
  <svg fill="none" {...svgProps} viewBox="0 0 66 66">
    {animations.map((animation) => (
      <circle
        key={animation.name || 'still'}
        cx="33"
        cy="33"
        fill="currentColor"
        r={animation.r * (animation.name ? (thickness / 100) : 1)}
        style={{
          opacity: animation.name ? 0.25 : 1,
          transformOrigin: 'center',
          animation: (!animation.name || still)
            ? 'none'
            : `${animation.name} ${140 / speed}s ease-in-out infinite`,
        }}
      />
    ))}
  </svg>
);

export const SpinnerRoundFilled = withSharedProps(Component);
