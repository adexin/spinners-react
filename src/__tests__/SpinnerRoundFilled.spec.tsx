import { render } from '@testing-library/react';
import React from 'react';

import { SpinnerRoundFilled } from '../SpinnerRoundFilled';

describe('SpinnerRoundFilled', () => {
  const color = 'red';
  const animations = [
    { r: 4 },
    {
      name: 'spinners-react-round-filled-inner',
      r: 12.66,
    },
    {
      name: 'spinners-react-round-filled-center',
      r: 20.32,
    },
    {
      name: 'spinners-react-round-filled-outer',
      r: 27.5,
    },
  ];

  it('matches the snapshot', () => {
    const { container } = render(
      <SpinnerRoundFilled color="#fff" size={100} speed={10} thickness={50} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('matches the snapshot with default props', () => {
    const { container } = render(<SpinnerRoundFilled />);

    expect(container).toMatchSnapshot();
  });

  it('renders null if disabled', () => {
    const { container } = render(<SpinnerRoundFilled enabled={false} />);

    expect(container.firstChild).toBe(null);
  });

  it('renders still if specified', () => {
    const { container } = render(<SpinnerRoundFilled still />);
    const circles = container.querySelectorAll('circle');

    circles.forEach((circle, index) => {
      expect(circle.style.animation).toBe('none');
      expect(circle).toHaveAttribute('r', `${animations[index].r}`);
    });
  });

  it('passes props to styles', () => {
    const size = '20%';
    const thickness = 40;
    const speed = 50;
    const { container } = render(
      <SpinnerRoundFilled
        color={color}
        size={size}
        speed={speed}
        thickness={thickness}
      />,
    );
    const circles = container.querySelectorAll('circle');

    expect(container.firstChild).toHaveStyle({ width: size });
    circles.forEach((circle, index) => {
      expect(
        circle.style.animation.includes(animations[index].name || 'none'),
      ).toBe(true);
      if (index) {
        expect(circle.style.animation.includes(`${140 / speed}`)).toBe(true);
      }
      expect(circle).toHaveAttribute(
        'r',
        `${
          index ? animations[index].r * (thickness / 100) : animations[index].r
        }`,
      );
    });
  });

  it('passes svg props overriding styles', () => {
    const className = 'test-class';
    const { container } = render(
      <SpinnerRoundFilled
        className={className}
        color="green"
        style={{ color }}
      />,
    );

    expect(container.firstChild).toHaveClass(className);
    expect(container.firstChild).toHaveStyle({ color });
  });
});
