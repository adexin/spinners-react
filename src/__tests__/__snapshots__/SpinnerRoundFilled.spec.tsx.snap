// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinnerRoundFilled matches the snapshot 1`] = `
<div>
  <svg
    fill="none"
    style="color: rgb(255, 255, 255); overflow: visible; width: 100px;"
    viewBox="0 0 66 66"
  >
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="4"
      style="opacity: 1; transform-origin: center; animation: none;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="6.33"
      style="opacity: 0.25; transform-origin: center; animation: spinners-react-round-filled-inner 14s ease-in-out infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="10.16"
      style="opacity: 0.25; transform-origin: center; animation: spinners-react-round-filled-center 14s ease-in-out infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="13.75"
      style="opacity: 0.25; transform-origin: center; animation: spinners-react-round-filled-outer 14s ease-in-out infinite;"
    />
  </svg>
</div>
`;

exports[`SpinnerRoundFilled matches the snapshot with default props 1`] = `
<div>
  <svg
    fill="none"
    style="color: rgb(56, 173, 72); overflow: visible; width: 50px;"
    viewBox="0 0 66 66"
  >
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="4"
      style="opacity: 1; transform-origin: center; animation: none;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="12.66"
      style="opacity: 0.25; transform-origin: center; animation: spinners-react-round-filled-inner 1.4s ease-in-out infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="20.32"
      style="opacity: 0.25; transform-origin: center; animation: spinners-react-round-filled-center 1.4s ease-in-out infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="27.5"
      style="opacity: 0.25; transform-origin: center; animation: spinners-react-round-filled-outer 1.4s ease-in-out infinite;"
    />
  </svg>
</div>
`;
